heat/ml/cluster/kmeans.py

```diff
@@ -4,51 +4,351 @@
+            Philadelphia, PA, USA. pp. 1027–1035, 2007.
+        """
+        self.init = init
         self.max_iter = max_iter
-        self.tol = tol
+        self.n_clusters = n_clusters
         self.random_state = random_state
+        self.tol = tol
+
+        # in-place properties
+        self._cluster_centers = None
+        self._labels = None
+        self._inertia = None
+        self._n_iter = None
+
+    @property
+    def cluster_centers_(self):
+        """
+        Returns
+        -------
+        ht.DNDarray, shape=(n_clusters, n_features):
+            Coordinates of cluster centers. If the algorithm stops before fully converging (see tol and max_iter),
+            these will not be consistent with labels_.
+        """
+        return self._cluster_centers.squeeze(axis=0).T
+
+    @property
+    def labels_(self):
+        """
+        Returns
+        -------
+        ht.DNDarray, shape=(n_points):
+            Labels of each point.
+        """
+        return self._labels
+
+    @property
+    def inertia_(self):
+        """
+        Returns
+        -------
+        float:
+            Sum of squared distances of samples to their closest cluster center.
+        """
+        return self._inertia
+
+    @property
+    def n_iter_(self):
+        """
+        Returns
+        -------
+        int:
+            Number of iterations run.
+        """
+        return self._n_iter
+
+    def _initialize_cluster_centers(self, X):
+        """
+        Initializes the K-Means centroids.
+
+        Parameters
+        ----------
+        X : ht.DNDarray, shape=(n_point, n_features)
+            The data to initialize the clusters for.
+        """
+        # always initialize the random state
+        if self.random_state is not None:
+            ht.random.seed(self.random_state)
+
+        # initialize the centroids by randomly picking some of the points
+        if self.init == 'random':
+            # Samples will be equally distributed drawn from all involved processes
+
+            _, displ, _ = X.comm.counts_displs_shape(shape=X.shape, axis=0)
+            centroids = ht.empty((1, X.shape[1], self.n_clusters), split=None)
+
+            if (X.split == None) or (X.split == 0):
+                for i in range(self.n_clusters):
+                    samplerange = (X.gshape[0] // self.n_clusters * i, X.gshape[0] // self.n_clusters * (i + 1))
+                    sample = ht.random.randint(samplerange[0], samplerange[1]).item()
+                    proc = 0
+                    for p in range(X.comm.size):
+                        if displ[p] > sample:
+                            break
+                        proc = p
+                    xi = ht.zeros(X.shape[1])
+                    if X.comm.rank == proc:
+                        idx = sample - displ[proc]
+                        xi = ht.array(X.lloc[idx, :, 0])
+                    xi.comm.Bcast(xi, root=proc)
```

Comment:
Can we do this with IBcasts and then simply wait for all of them at the end? Would be slightly faster I presume.

Reply:
done